r2k/r2k/source/main.c

```diff
@@ -0,0 +1,502 @@
+				Status = STATUS_BUFFER_TOO_SMALL;
+				break;
+			}
+
+
+			pMods = (PRTL_PROCESS_MODULES)ExAllocatePoolWithTag(NonPagedPool, bytes, 'domP');
+			RtlZeroMemory(pMods, bytes);
+
+			Status = ZwQuerySystemInformation(SystemModuleInformation, pMods, bytes, &bytes);
+			RtlCopyMemory(pOutBuf, (void*)pMods, bytes);
+			if (pMods)
+				ExFreePoolWithTag(pMods, 'domP');
+			dwBytesWritten = bytes;
+			Status = STATUS_SUCCESS;
+			break;
+		}
+		case IOCTL_READ_KERNEL_MEM:
+		{
+			
+			UINT32 len = 0;
+			LARGE_INTEGER virt_addr = { 0x0, 0x0 };
+			if (!Irp->AssociatedIrp.SystemBuffer || IrpSp->Parameters.DeviceIoControl.InputBufferLength < 3 * sizeof(UINT32))
+			{
+				DbgPrint("[R2K] IOCTL_READ_KERNEL_MEM ERROR: STATUS_INVALID_PARAMETER\n");
+				Status = STATUS_INVALID_PARAMETER;
+				break;
+			}
+
+			pInBuf = Irp->AssociatedIrp.SystemBuffer;
+			pOutBuf = Irp->AssociatedIrp.SystemBuffer;
+
+			virt_addr.LowPart	= ((UINT32*)pInBuf)[0];
+			virt_addr.HighPart	= ((UINT32*)pInBuf)[1];
+			len					= ((UINT32*)pInBuf)[2];
+			if (!len) len = 4;
+
+			if (IrpSp->Parameters.DeviceIoControl.OutputBufferLength < len)
+			{
+				DbgPrint("[R2K] IOCTL_READ_KERNEL_MEM  ERROR: STATUS_BUFFER_TOO_SMALL\n");
+				Status = STATUS_BUFFER_TOO_SMALL;
+				break;
+			}
+
+			__try
+			{
+				//PAGE_SIZE
+				// is addres + len valid page?
+				if (!MmIsAddressValid((void*)(LONG_PTR)(virt_addr.QuadPart + len)))
+				{
+					DbgPrint("[R2K] IOCTL_READ_KERNEL_MEM: Error page not valid at addres + len %p\n", (virt_addr.QuadPart + len));
+					Status = STATUS_ACCESS_DENIED;
+					break;
+				}
+				if (MmIsAddressValid((void*)(LONG_PTR)virt_addr.QuadPart))
+				{
+					DbgPrint("[R2K] IOCTL_READ_KERNEL_MEM: Reading address: %p\n", virt_addr.QuadPart);
+					DbgPrint("                             Bytes to read  : %u\n", len);
+					RtlCopyMemory(pOutBuf, (void*)(LONG_PTR)virt_addr.QuadPart, len);
+					Status = STATUS_SUCCESS;
+				}
+				else
+				{
+					DbgPrint("[R2K] IOCTL_READ_KERNEL_MEM: Error page not valid %p\n", virt_addr.QuadPart);
+					Status = STATUS_ACCESS_DENIED;
+					break;
+				}
+			}
+			__except (EXCEPTION_EXECUTE_HANDLER)
+			{
+				Status = GetExceptionCode();
+				DbgPrint("[R2K] IOCTL_READ_KERNEL_MEM ERROR: exception code 0x%X\n", Status);
+				break;
+			}
+
+			if (NT_SUCCESS(Status))
+			{
+				dwBytesWritten = len;
+			}
+			break;
+		}
+		case IOCTL_WRITE_KERNEL_MEM:
+		{
+			UINT32 len = 0;
+			LARGE_INTEGER virt_addr = { 0x0, 0x0 };
+			unsigned char* 		buffer = 0;
+			if (!Irp->AssociatedIrp.SystemBuffer || IrpSp->Parameters.DeviceIoControl.InputBufferLength < 4 * sizeof(UINT32))
+			{
+				DbgPrint("[R2K] IOCTL_WRITE_KERNEL_MEM ERROR: STATUS_INVALID_PARAMETER\n");
+				Status = STATUS_INVALID_PARAMETER;
+				break;
+			}
+
+			pInBuf = Irp->AssociatedIrp.SystemBuffer;
+			pOutBuf = Irp->AssociatedIrp.SystemBuffer;
+
+			virt_addr.LowPart	= ((UINT32*)pInBuf)[0];
```

Comment:
do not use tabs if its not for indenting, align with spaces or just dont align :)